### PR TITLE
[sidelining] simplify algorithm by removing specialization

### LINKED ISSUE #961: simplify code for no-specialization PR, remove multipass, rerun experiments

### PR DESCRIPTION
**Pull request by NishanthJKumar**
Gets rid of specialization in favor of always spawning.

Supercloud results from before this change: https://github.com/Learning-and-Intelligent-Systems/predicators/pull/931

Supercloud results (only oracle and backchaining) from after this change:
TODO

closes #961

### CODE REVIEW

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 1]
```diff
@@ -85,6 +86,11 @@ def _learn(self) -> List[PartialNSRTAndDatastore]:
             # Run one pass of backchaining.
             nec_pnad_set_changed = self._backchain_one_pass(
                 param_opt_to_nec_pnads, param_opt_to_general_pnad)
+            num_backchaining_iters += 1
+        
+        # Test to check that there are only ever two iterations of backchaining
+        # that get run.
+        assert num_backchaining_iters == 2
```
Comment (tomsilver) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:93]: why 2 and not 1?

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 2]
```diff
@@ -145,45 +151,39 @@ def _backchain_one_pass(
...
-                    # If we were unable to specialize any of the PNADs, we need
-                    # to spawn from the most general PNAD and make a new PNAD
-                    # to cover these necessary add effects.
-                    else:
-                        new_pnad = self._try_specializing_pnad(
-                            necessary_add_effects,
-                            param_opt_to_general_pnad[option.parent], segment)
-                        assert new_pnad is not None
-
+                    new_pnad = self._try_specializing_pnad(
```
Comment (tomsilver) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:159]: I'm guessing this will go away with #961 but looks okay for now

**Review by tomsilver — COMMENTED** [review]
exciting times!

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py (hunk 2 above)
Comment (NishanthJKumar) [reply src/nsrt_learning/strips_learning/gen_to_spec_learner.py]: yup - it has since been removed

**Review by ronuchit — APPROVED** [review]

**Review by ronuchit — APPROVED** [review]

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 3]
```diff
@@ -144,46 +130,30 @@ def _backchain_one_pass(
...
                     # matched to another PNAD.
                     self._recompute_datastores_from_segments(
                         param_opt_to_nec_pnads[option.parent])
-                    # Recompute all preconditions, now that we have recomputed
-                    # the datastores.
-                    for nec_pnad in param_opt_to_nec_pnads[option.parent]:
-                        pre = self._induce_preconditions_via_intersection(
-                            nec_pnad)
-                        nec_pnad.op = nec_pnad.op.copy_with(preconditions=pre)
+                    # If any PNAD now has an empty datastore, delete it. Then
```
Comment (tomsilver) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:148]: just wondering, is this change about deleting empty datastores important? or just nice to have?

**Review by tomsilver — APPROVED** [review]

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py (hunk 3 above)
Comment (ronuchit) [reply src/nsrt_learning/strips_learning/gen_to_spec_learner.py:165]: making this a todo for later

**Comment by NishanthJKumar** [comment]
Subsumed by #976

**Pull request closed without merging**